research/conditional/moe_layers/token_choice.py

```diff
@@ -81,57 +87,84 @@ def forward(self, x: torch.Tensor):
             expert_mask = expanded_expert_mask.sum(dim=1)
             assert expert_mask.shape == (n_tokens, self.n_experts)
 
-        with measure_time(self, "calculate expert indexes"):
-            position_in_expert = torch.cumsum(expert_mask, dim=0) * expert_mask
-            in_capacity_tokens_mask = torch.lt(position_in_expert, capacity)
-            expert_mask *= in_capacity_tokens_mask
-            n_selected_tokens = expert_mask.sum().item()
-
-            self.update_cache_for_logging(
-                "dropped_tokens_ratio",
-                ((n_tokens * self.routing_top_k) - n_selected_tokens)
-                / (n_tokens * self.routing_top_k),
-            )
+        if self.vectorize:
+            with measure_time(self, "experts_lists"):
+                (
+                    top_tokens_per_expert_values,
+                    top_tokens_per_expert_indices,
+                ) = expert_mask.topk(k=capacity, dim=0)
```

Comment:
`topk` isn't stable as per https://github.com/pytorch/pytorch/pull/88619
We are changing behavior to dropping random tokens

Reply:
Yes